### PR TITLE
Add reactive power control mode to ratio tap changers

### PR DESCRIPTION
**Pull request by p-arvy**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

No

**What kind of change does this PR introduce?**

Feature: Add a regulation mode "reactive power" to ratio tap changers

**What is the current behavior?**

Ratio tap changers can only regulate voltage

**What is the new behavior (if this is a feature change)?**
Ratio tap changers can regulate either voltage or reactive power

**Does this PR introduce a breaking change or deprecate an API?**

No, iIDM API change and version bump to v1.12, but backward compatible as follows:

- RTC `regulationMode` is introduced, with possible values `VOLTAGE` or `REACTIVE_POWER`. `getRegulationMode` / `setRegulationMode`
- `targetV` becomes `regulationValue`, `getRegulationValue` / `setRegulationValue`
- backward compatibility ensured by keeping `getTargetV` and `setTargetV`, acting as shortcuts as follows:
  - setTargetV automatically switches mode to VOLTAGE (if new targetV is not NaN)
  - getTargetV returns NaN if mode is not VOLTAGE

**Other information**:

### CODE REVIEW

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java [hunk 1]
```diff
@@ -13,20 +13,52 @@
  */
 public interface RatioTapChanger extends TapChanger<RatioTapChanger, RatioTapChangerStep> {
 
+    enum RegulationMode {
+        VOLTAGE,
+        REACTIVE_POWER_CONTROL
```
Comment (annetill) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java:18]: REACTIVE_POWER here.

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java [hunk 2]
```diff
@@ -745,6 +745,7 @@ private static boolean addControlThreeWindingsTransformerLeg(Network network, St
 
     private static boolean defineVoltageControl(Network network, String id, PsseTransformerWinding winding, RatioTapChanger rtc,
         boolean regulatingForcedToOff) {
+        // TODO : remove this ?
```
Comment (Hadrien-Godard) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java:748]: I think you could remove this test and only keep the "if (Math.abs(winding.getCod()) != 1)"

File: ucte/ucte-converter/src/main/java/com/powsybl/ucte/converter/UcteExporter.java [hunk 3]
```diff
@@ -722,8 +722,10 @@ private static UctePhaseRegulation convertRatioTapChanger(TwoWindingsTransformer
                 twoWindingsTransformer.getRatioTapChanger().getHighTapPosition(),
                 twoWindingsTransformer.getRatioTapChanger().getTapPosition(),
                 Double.NaN);
-        if (!Double.isNaN(twoWindingsTransformer.getRatioTapChanger().getTargetV())) {
-            uctePhaseRegulation.setU(twoWindingsTransformer.getRatioTapChanger().getTargetV());
+        // TODO : add reactive power case ?
```
Comment (Hadrien-Godard) [inline ucte/ucte-converter/src/main/java/com/powsybl/ucte/converter/UcteExporter.java:725]: If UCTE format supports it, yes. If not, no.

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/CgmesExportUtil.java [hunk 4]
```diff
@@ -383,7 +383,8 @@ private static <C extends Connectable<C>> String getTapChangerId(C twt, String c
     }
 
     static boolean regulatingControlIsDefined(RatioTapChanger rtc) {
-        return !Double.isNaN(rtc.getTargetV())
```
Comment (annetill) [inline cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/CgmesExportUtil.java:386]: Here, I prefer that you keep previous code because reactive power control is not yet supported in CGMES conversion.

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/SteadyStateHypothesisExport.java [hunk 5]
```diff
@@ -411,7 +411,7 @@ private static void addRegulatingControlView(TapChanger<?, ?> tc, CgmesTapChange
                         true,
                         ratioTapChanger.isRegulating(),
                         ratioTapChanger.getTargetDeadband(),
-                        ratioTapChanger.getTargetV(),
```
Comment (annetill) [inline cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/SteadyStateHypothesisExport.java:414]: Same, keep targetV.

File: cgmes/cgmes-conversion/src/test/java/com/powsybl/cgmes/conversion/test/network/compare/Comparison.java [hunk 6]
```diff
@@ -715,7 +715,12 @@ private void compareRatioTapChanger(
         compare("ratioTapChanger.hasLoadTapChangingCapabilities",
                 expected.hasLoadTapChangingCapabilities(),
                 actual.hasLoadTapChangingCapabilities());
-        compare("ratioTapChanger.targetV", expected.getTargetV(), actual.getTargetV());
```
Comment (annetill) [inline cgmes/cgmes-conversion/src/test/java/com/powsybl/cgmes/conversion/test/network/compare/Comparison.java:718]: Keep previous.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java [hunk 7]
```diff
@@ -13,20 +13,52 @@
  */
 public interface RatioTapChanger extends TapChanger<RatioTapChanger, RatioTapChangerStep> {
 
+    enum RegulationMode {
+        VOLTAGE,
+        REACTIVE_POWER
+    }
+
+    /**
+     * Get the regulation mode.
```
Comment (annetill) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java:22]: Can you increase the javadoc here with supported modes.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java [hunk 8]
```diff
@@ -13,20 +13,52 @@
...
+    /**
+     * Set the regulation mode
+     * @param regulationMode the regulation mode
+     * @return itself for method chaining
+     */
+    RatioTapChanger setRegulationMode(RatioTapChanger.RegulationMode regulationMode);
+
     /**
-     * Get the target voltage in kV.
+     * Get the regulation value.
```
Comment (annetill) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java:35]: Same, more details.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/ValidationUtil.java [hunk 9]
```diff
@@ -479,25 +479,31 @@ private static ValidationLevel errorOrWarningForRtc(Validable validable, boolean
...
     public static ValidationLevel checkRatioTapChangerRegulation(Validable validable, boolean regulating, boolean loadTapChangingCapabilities,
-                                                                 Terminal regulationTerminal, double targetV, Network network, boolean throwException,
+                                                                 Terminal regulationTerminal, RatioTapChanger.RegulationMode regulationMode,
+                                                                 double regulationValue, Network network, boolean throwException,
                                                                  Reporter reporter) {
         ValidationLevel validationLevel = ValidationLevel.STEADY_STATE_HYPOTHESIS;
-        if (regulating) {
-            if (Double.isNaN(targetV)) {
-                validationLevel = ValidationLevel.min(validationLevel, errorOrWarningForRtc(validable, loadTapChangingCapabilities, "a target voltage has to be set for a regulating ratio tap changer", throwException, reporter));
+        if (regulationMode == null) {
```
Comment (annetill) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/ValidationUtil.java:498]: If `regulating`, a regulation mode is needed, but if no regulation, why do you have this treatment?

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerImpl.java [hunk 10]
```diff
@@ -97,27 +101,63 @@ public RatioTapChangerImpl setLoadTapChangingCapabilities(boolean loadTapChangin
 
     @Override
     public double getTargetV() {
-        return targetV.get(network.get().getVariantIndex());
+        if (regulationMode != RegulationMode.VOLTAGE) {
+            throw new IllegalAccessError("Regulation mode must be in voltage to access to target V");
```
Comment (annetill) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerImpl.java:105]: Use a constant for message.

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java [hunk 11]
```diff
@@ -82,6 +82,7 @@ public enum ConnectedComponentMode {
 
     public static final VoltageInitMode DEFAULT_VOLTAGE_INIT_MODE = VoltageInitMode.UNIFORM_VALUES;
     public static final boolean DEFAULT_TRANSFORMER_VOLTAGE_CONTROL_ON = false;
+    public static final boolean DEFAULT_TRANSFORMER_REACTIVE_POWER_CONTROL_ON = false;
```
Comment (annetill) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java:85]: Not in the same PR...

**Review by annetill — CHANGES_REQUESTED** [review]

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/TransformersValidation.java [hunk 12]
```diff
@@ -79,7 +79,7 @@ public boolean checkTransformer(TwoWindingsTransformer twt, ValidationConfig con
         double rho = ratioTapChanger.getCurrentStep().getRho();
         double rhoPreviousStep = tapPosition == lowTapPosition ? Double.NaN : ratioTapChanger.getStep(tapPosition - 1).getRho();
         double rhoNextStep = tapPosition == highTapPosition ? Double.NaN : ratioTapChanger.getStep(tapPosition + 1).getRho();
-        double targetV = ratioTapChanger.getTargetV();
+        double regulationValue = ratioTapChanger.getRegulationValue();
```
Comment (jeandemanged) [inline loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/TransformersValidation.java:82]: revert to using targetV variable name, set to NaN if regulation mode is not voltage

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/TransformersValidation.java [hunk 13]
```diff
@@ -90,7 +90,7 @@ public boolean checkTransformer(TwoWindingsTransformer twt, ValidationConfig con
                         ValidationType.TWTS, ValidationUtils.VALIDATION_WARNING, twt.getId());
             try {
                 twtsWriter.write(twt.getId(), Float.NaN, Float.NaN, Float.NaN, rho, rhoPreviousStep, rhoNextStep, tapPosition, lowTapPosition,
-                                 highTapPosition, targetV, null, Float.NaN, false, false, true);
+                                 highTapPosition, regulationValue, null, Float.NaN, false, false, true);
```
Comment (jeandemanged) [inline loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/TransformersValidation.java:93]: revert to using targetV variable name

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/TransformersValidation.java [hunk 14]
```diff
@@ -103,7 +103,7 @@ public boolean checkTransformer(TwoWindingsTransformer twt, ValidationConfig con
         boolean connectableMainComponent = connectableBus != null && connectableBus.isInMainConnectedComponent();
         boolean mainComponent = bus != null ? bus.isInMainConnectedComponent() : connectableMainComponent;
         return checkTransformer(twt.getId(), rho, rhoPreviousStep, rhoNextStep, tapPosition, lowTapPosition, highTapPosition,
-                                 targetV, regulatedSide, v, connected, mainComponent, config, twtsWriter);
+                regulationValue, regulatedSide, v, connected, mainComponent, config, twtsWriter);
```
Comment (jeandemanged) [inline loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/TransformersValidation.java:106]: revert to using targetV variable name

File: loadflow/loadflow-validation/src/test/java/com/powsybl/loadflow/validation/TransformersValidationTest.java [hunk 15]
```diff
@@ -81,7 +81,7 @@ void setUp() throws IOException {
         Mockito.when(ratioTapChanger.getCurrentStep()).thenReturn(currentStep);
         Mockito.when(ratioTapChanger.getStep(tapPosition - 1)).thenReturn(previousStep);
         Mockito.when(ratioTapChanger.getStep(tapPosition + 1)).thenReturn(nextStep);
-        Mockito.when(ratioTapChanger.getTargetV()).thenReturn(targetV);
+        Mockito.when(ratioTapChanger.getRegulationValue()).thenReturn(targetV);
```
Comment (jeandemanged) [inline loadflow/loadflow-validation/src/test/java/com/powsybl/loadflow/validation/TransformersValidationTest.java:84]: revert to using targetV variable name

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java (hunk 2 above)
Comment (jeandemanged) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java:748]: remove this todo, will be in the scope of another PR

File: ucte/ucte-converter/src/main/java/com/powsybl/ucte/converter/UcteExporter.java (hunk 3 above)
Comment (jeandemanged) [reply ucte/ucte-converter/src/main/java/com/powsybl/ucte/converter/UcteExporter.java:725]: UCTE format does not support reactive power control mode, remove the TODO

File: ucte/ucte-converter/src/main/java/com/powsybl/ucte/converter/UcteImporter.java [hunk 16]
```diff
@@ -500,10 +500,12 @@ private static RatioTapChangerAdder createRatioTapChangerAdder(UctePhaseRegulati
                 .setLowTapPosition(lowerTap)
                 .setTapPosition(uctePhaseRegulation.getNp())
                 .setLoadTapChangingCapabilities(!Double.isNaN(uctePhaseRegulation.getU()));
+        // TODO : add reactive power case ?
```
Comment (jeandemanged) [inline ucte/ucte-converter/src/main/java/com/powsybl/ucte/converter/UcteImporter.java:503]: UCTE format does not support reactive power control mode, remove the TODO

File: iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/AbstractTransformerSerDe.java [hunk 17]
```diff
@@ -79,7 +79,7 @@ protected static void writeRatioTapChanger(String name, RatioTapChanger rtc, Net
         }
         writeTapChanger(rtc, context);
         context.getWriter().writeBooleanAttribute("loadTapChangingCapabilities", rtc.hasLoadTapChangingCapabilities());
-        context.getWriter().writeDoubleAttribute("targetV", rtc.getTargetV());
+        context.getWriter().writeDoubleAttribute("targetV", rtc.getRegulationValue());
```
Comment (jeandemanged) [inline iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/AbstractTransformerSerDe.java:82]: iIDM SerDe to be updated for backward compatibility for iIDM version < 1.12

**Review by jeandemanged — CHANGES_REQUESTED** [review]
mainly iIDM ser/deser-ialization to be fixed with backward compatibility support

some cleanup to do

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java [hunk 18]
```diff
@@ -772,14 +773,17 @@ private static boolean defineVoltageControl(Network network, String id, PsseTran
...
+        rtc.setRegulationMode(RatioTapChanger.RegulationMode.VOLTAGE)
+            .setRegulationValue(targetV)
             .setTargetDeadband(targetDeadBand)
             .setRegulationTerminal(regulatingTerminal)
             .setRegulating(regulating);
 
         return regulating;
     }
 
+    // TODO : add reactive power control
```
[the remark this answers is not in the document]
Comment (jeandemanged) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java]: psse not in scope o this PR

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java (hunk 2 above)
Comment (jeandemanged) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java:748]: psse would be done in another PR

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java [hunk 19]
```diff
@@ -13,20 +13,63 @@
...
     /**
      * Get the target voltage in kV.
-     * <p>
-     * Depends on the working variant.
+     * <p>Needed only when regulating mode is set to {@link RegulationMode#VOLTAGE}.</p>
+     * <p>Depends on the working variant.</p>
      * @see VariantManager
+     * @deprecated Use {@link RatioTapChanger#getRegulationValue()} instead.
      */
+    @Deprecated(since = "6.1.0")
```
[the remark this answers is not in the document]
Comment (p-arvy) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java]: my bad, wrong handling during rebase

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerAdderImpl.java [hunk 20]
```diff
@@ -144,7 +146,22 @@ public RatioTapChangerAdder setRegulating(boolean regulating) {
...
+        if (this.regulationMode != RatioTapChanger.RegulationMode.VOLTAGE) {
+            throw new IllegalAccessError("Regulation mode must be in voltage to set target V");
+        }
+        this.regulationValue = targetV;
+        return this;
+    }
+
+    @Override
+    public RatioTapChangerAdder setRegulationMode(RatioTapChanger.RegulationMode regulationMode) {
+        this.regulationMode = regulationMode;
```
[the remark this answers is not in the document]
Comment (p-arvy) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerAdderImpl.java:158]: If so, no null value possible. Regulation mode must always be given, even if no regulation. Do we want to?

File: iidm/iidm-serde/src/main/resources/xsd/iidm_V1_12.xsd [hunk 21]
```diff
@@ -535,7 +541,8 @@
         <xs:attribute name="targetDeadband" use="optional" type="xs:double"/>
         <xs:attribute name="loadTapChangingCapabilities" use="required" type="xs:boolean"/>
         <xs:attribute name="regulating" use="optional" type="xs:boolean"/>
-        <xs:attribute name="targetV" use="optional" type="xs:double"/>
+        <xs:attribute name="regulationMode" use="required" type="iidm:RatioRegulationMode"/>
```
Comment (p-arvy) [inline iidm/iidm-serde/src/main/resources/xsd/iidm_V1_12.xsd:544]: Should the regulationMode be required there @annetill @jeandemanged ?

File: iidm/iidm-serde/src/main/resources/xsd/iidm_equipment_V1_12.xsd [hunk 22]
```diff
@@ -531,7 +537,8 @@
         <xs:attribute name="targetDeadband" use="optional" type="xs:double"/>
         <xs:attribute name="loadTapChangingCapabilities" use="required" type="xs:boolean"/>
         <xs:attribute name="regulating" use="optional" type="xs:boolean"/>
-        <xs:attribute name="targetV" use="optional" type="xs:double"/>
+        <xs:attribute name="regulationMode" use="required" type="iidm:RatioRegulationMode"/>
```
Comment (p-arvy) [inline iidm/iidm-serde/src/main/resources/xsd/iidm_equipment_V1_12.xsd:540]: Same here, should be required @annetill @jeandemanged  ?

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerImpl.java (hunk 10 above)
Comment (jeandemanged) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerImpl.java]: message not needed anymore, changed to return Double.NaN if regulation mode is not voltage

**Review by jeandemanged — APPROVED** [review]

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java [hunk 23]
```diff
@@ -7,25 +7,69 @@
 package com.powsybl.iidm.network;
 
 /**
- * A ratio tap changer that is associated to a transformer to control the voltage.
+ * A ratio tap changer that is associated to a transformer to control the ratio.
```
Comment (annetill) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java:10]: The ratio?

Comment (p-arvy) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java:10]: This formulation to be consistent with PhaseTapChanger javadoc. 
Change to "... to control the voltage or reactive power"?

Comment (annetill) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChanger.java:10]: I think there is an issue in phase tap changer javadoc.

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerAdderImpl.java [hunk 24]
```diff
@@ -144,7 +143,22 @@ public RatioTapChangerAdder setRegulating(boolean regulating) {
 
     @Override
     public RatioTapChangerAdder setTargetV(double targetV) {
-        this.targetV = targetV;
+        if (!Double.isNaN(targetV)) {
```
Comment (annetill) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerAdderImpl.java:146]: I think @jeandemanged you just have to check if regulation mode is different than VOLTAGE, no?

**Review by annetill — CHANGES_REQUESTED** [review]
Small comments. I don't see any file with a REACTIVE_POWER mode serialisation.

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerAdderImpl.java (hunk 24 above)
Comment (jeandemanged) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerAdderImpl.java:146]: hm, checking for what purpose ?
if API user sets a targetV with a non-NaN value, then implicitly he wants to be in VOLTAGE regulation mode

Comment (annetill) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/RatioTapChangerAdderImpl.java:146]: Because I understand that you can have a NaN targetV and have a REACTIVE_POWER mode, but set a targetV that will change the mode, no?

File: iidm/iidm-serde/src/test/java/com/powsybl/iidm/serde/RatioTapChangerSerDeTest.java [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+                .beginStep()
+                .setRho(0.9)
+                .setR(0.1089)
+                .setX(0.01089)
+                .setG(0.8264462809917356)
+                .setB(0.08264462809917356)
+                .endStep()
+                .setTapPosition(0)
+                .setLoadTapChangingCapabilities(true)
+                .setRegulating(false)
```
Comment (annetill) [inline iidm/iidm-serde/src/test/java/com/powsybl/iidm/serde/RatioTapChangerSerDeTest.java:81]: @p-arvy Can you put regulating to `true` just to be sure that a regulating terminal is asked.

Comment (p-arvy) [reply iidm/iidm-serde/src/test/java/com/powsybl/iidm/serde/RatioTapChangerSerDeTest.java]: @annetill done

**Review by annetill — APPROVED** [review]
